### PR TITLE
Expose transfer claims on CCLF JSON API

### PR DESCRIPTION
**Pull request by jsugarman**
#### What
Expose transfer claims on CCLF JSON API

#### Why
To enable injection of Transfer claims and there "bills"
into CCLF.

#### Ticket
[pivotal ticket](https://www.pivotaltracker.com/story/show/155510029)

#### How
Add bill scenarios to the "Transfer Brain" and call these 
from the JSON API adapters for Transfer claims.

--------

#### TODO (wip)

 - [X] refactor transfer brain 
 - [ ] refactor transfer brain data item class
 - [X] add transfer claim entity
 - [X] add transfer fee adapter
 - [x] map transfer fee bills
 - [x] map all other transfer claim bills (disbursements/expense, misc fees)

### CODE REVIEW

File: spec/models/claim/transfer_brain_data_item_spec.rb [hunk 1]
```diff
@@ -3,60 +3,75 @@
 module Claim
   describe TransferBrainDataItem do
```
Comment (jsugarman) [inline spec/models/claim/transfer_brain_data_item_spec.rb]: Change to `RSpec.describe Claim::TransferBrainItem`

File: spec/support/matchers/hash_matchers.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    msg += "Diff: (see HashDiff - https://github.com/liufengyun/hashdiff):\n"
+    msg += format_diff(@diff)
+    msg
+  end
+
+  failure_message_when_negated do |owner|
+    "expected hashes not to match"
+  end
+
+  # TODO: colorize the 3rd element as success/green to indicate this is what was received
```
Comment (jsugarman) [inline spec/support/matchers/hash_matchers.rb]: remove as more complicated for nils

File: spec/support/transfer_brain_helpers.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+           30=>{"*"=>{ :validity=>true, :transfer_fee_full_name=>"during trial transfer (org) - trial", :allocation_type=>"Grad", :bill_scenario=>'TODO'}},
+           40=>{"*"=>{ :validity=>true, :transfer_fee_full_name=>"transfer after trial and before sentence hearing (org)", :allocation_type=>"Grad", :bill_scenario=>'TODO'}},
+           50=>{"*"=>{ :validity=>true, :transfer_fee_full_name=>"transfer before retrial (org) - retrial", :allocation_type=>"Grad", :bill_scenario=>'TODO'}},
+           60=>{"*"=>{ :validity=>true, :transfer_fee_full_name=>"transfer during retrial (org) - retrial", :allocation_type=>"Grad", :bill_scenario=>'TODO'}},
+           70=>{"*"=>{ :validity=>true, :transfer_fee_full_name=>"transfer after retrial and before sentence hearing (org)", :allocation_type=>"Grad", :bill_scenario=>'TODO'}}
+          }
+       }
+    }
+  end
+end
```
Comment (jsugarman) [inline spec/support/transfer_brain_helpers.rb]: line break

File: app/models/claim/transfer_brain_data_item.rb [hunk 4]
```diff
@@ -1,24 +1,26 @@
...
+      @bill_scenario          = data_item.bill_scenario
     end
 
-    def match_detail?(detail)
-      @litigator_type == detail.litigator_type &&
-        @elected_case == detail.elected_case &&
-        @transfer_stage_id == detail.transfer_stage_id &&
-        @case_conclusion_id == detail.case_conclusion_id
-    end
+    # def match_detail?(detail)
```
Comment (jsugarman) [inline app/models/claim/transfer_brain_data_item.rb]: remove

File: app/models/claim/transfer_brain_data_item.rb [hunk 5]
```diff
@@ -1,23 +1,18 @@
 module Claim
   class TransferBrainDataItem
-    attr_reader :litigator_type, :elected_case, :transfer_stage_id, :case_conclusion_id,
-                :allocation_type, :validity, :transfer_fee_full_name
+    # FIXME: does not look like these are used anywhere
+    # attr_reader :litigator_type, :elected_case, :transfer_stage_id, :case_conclusion_id,
+    # :allocation_type, :validity, :transfer_fee_full_name, :bill_scenario
```
Comment (colinbruce) [inline app/models/claim/transfer_brain_data_item.rb]: Can this be deleted if they aren't used?

**Review by colinbruce — APPROVED** [review]
1 minor observation, but otherwise 👍

File: app/models/claim/transfer_brain_data_item.rb (hunk 5 above)
Comment (jsugarman) [reply app/models/claim/transfer_brain_data_item.rb]: hopefully, yes. I am going to do more of a refactor on this file/functionality anyway so will do it then.

File: app/models/claim/transfer_brain_data_item_collection.rb [hunk 6]
```diff
@@ -92,16 +95,33 @@ def valid_case_conclusion_ids(litigator_type, elected_case, transfer_stage_id)
...
+      end
+    end
+
+    def load_data_items
+      csv = read_csv
+      attributes = csv.headers.map(&:to_sym)
+      klass = Struct.new('DataItem', *attributes)
+
+      data_items = csv.each_with_object([]) do |row, arr|
+        arr << klass.new(*row.to_hash.values)
```
Comment (jsugarman) [inline app/models/claim/transfer_brain_data_item_collection.rb:121]: try parsing to transfer brain item here

File: app/services/cclf/case_type_adapter.rb [hunk 7]
```diff
@@ -35,20 +35,29 @@ def initialize(claim)
...
 
     def interim_bill_scenario
       BILL_SCENARIOS[claim.interim_fee&.fee_type&.unique_code&.to_sym]
     end
 
+    def transfer_scenario_applicable?
+      claim.transfer? && transfer_bill_scenario
+    end
+
+    # TODO: transfer claim bill scenarios - not all mapped in transfer_brain_data_items.csv
```
Comment (jsugarman) [inline app/services/cclf/case_type_adapter.rb]: remove

File: lib/extensions/hash_extension.rb [hunk 8]
```diff
@@ -5,4 +5,11 @@ def all_keys
       keys.concat(v.all_keys) if v.is_a? Hash
     end
   end
+
+  def all_values_for(key)
+    each_with_object([]) do |(k, v), result|
+      result << v if k.eql?(key)
```
Comment (jsugarman) [inline lib/extensions/hash_extension.rb]: add line for arrays containing hashes

File: app/interfaces/api/entities/cclf/transfer_claim.rb [hunk 9]
```diff
@@ -5,11 +5,10 @@ class TransferClaim < BaseClaim
         # TODO: WIP - all bills must be addeded
```
Comment (jsugarman) [inline app/interfaces/api/entities/cclf/transfer_claim.rb]: remove

File: spec/api/entities/cclf/case_type_spec.rb [hunk 10]
```diff
@@ -4,20 +4,41 @@
...
 
+  context 'delegation' do
+    let(:claim) { instance_double(::Claim::LitigatorClaim, case_type: case_type, interim?: false, transfer?: false) }
+    let(:adapter_klass) { ::CCLF::CaseTypeAdapter }
+    let(:adapter) { instance_double(adapter_klass) }
+
+    it "delegates bill_scenario to adapter" do
+      expect(adapter_klass).to receive(:new).with(claim).and_return(adapter)
+      expect(adapter).to receive(:bill_scenario)
+      subject
```
Comment (jsugarman) [inline spec/api/entities/cclf/case_type_spec.rb]: change to response for clarity

File: spec/api/v2/cclf_claim_spec.rb [hunk 11]
```diff
@@ -56,6 +63,18 @@
   end
 end
 
+RSpec.shared_examples 'litigator fee bill' do
+  it 'returns array containing a litigator fee bill' do
+    is_expected.to have_json_size(1).at_path("bills")
+  end
+
+  it 'returns a litigator fee bill' do
+    is_expected.to have_json_size(1).at_path("bills")
```
Comment (jsugarman) [inline spec/api/v2/cclf_claim_spec.rb]: remove this line as separate test above

File: spec/api/v2/cclf_claim_spec.rb [hunk 12]
```diff
@@ -56,6 +63,18 @@
   end
 end
 
+RSpec.shared_examples 'litigator fee bill' do
+  it 'returns array containing a litigator fee bill' do
```
Comment (jsugarman) [inline spec/api/v2/cclf_claim_spec.rb]: change `returns array containing 1 bill`

File: spec/factories/fee_types.rb [hunk 13]
```diff
@@ -244,12 +244,13 @@
     factory :transfer_fee_type, class: Fee::TransferFeeType do
       calculated false
       code 'TRANS'
+      # unique_code 'TRANS' # TODO - add and fix broken specs
```
Comment (jsugarman) [inline spec/factories/fee_types.rb]: add this and get broken spec (resulting from it) passing

File: spec/support/transfer_brain_helpers.rb [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+  class_methods do
+    def data_item_collection_hash
+      DATA_ITEM_COLLECTION_HASH
+    end
+
+    def transfer_fee_bill_scenarios
+      data_item_collection_hash.
+        all_values_for(:bill_scenario).
+        reject { |scenario| scenario.nil? }
```
Comment (jsugarman) [inline spec/support/transfer_brain_helpers.rb]: change to `reject(&:nil?)`

File: spec/validators/claim/transfer_claim_validator_spec.rb [hunk 15]
```diff
@@ -92,7 +92,7 @@
 
   context 'trial dates validation' do
     context 'case type: trial' do
-      let(:claim) { build :transfer_claim, :trial }
+      let(:claim) { build :transfer_claim, :with_transfer_detail }
```
Comment (jsugarman) [inline spec/validators/claim/transfer_claim_validator_spec.rb]: add braces

File: spec/validators/claim/transfer_claim_validator_spec.rb [hunk 16]
```diff
@@ -103,7 +103,7 @@
       end
     end
     context 'case type: retrial' do
-      let(:claim) { build :transfer_claim, :retrial }
+      let(:claim) { build :transfer_claim, :with_transfer_detail }
```
Comment (jsugarman) [inline spec/validators/claim/transfer_claim_validator_spec.rb]: add braces

File: spec/validators/claim/transfer_claim_validator_spec.rb [hunk 17]
```diff
@@ -115,6 +115,18 @@
     end
   end
 
+  context 'case type validation' do
+     let(:claim) { build :transfer_claim, :with_transfer_detail }
```
Comment (jsugarman) [inline spec/validators/claim/transfer_claim_validator_spec.rb]: add braces

**Review by jsugarman — COMMENTED** [review]
self review

**Review by lostie — APPROVED** [review]
codewise, LGTM 👍 .

**Pull request merged**